### PR TITLE
Add an associated entity instruction and corresponding type for interface elements.

### PR DESCRIPTION
**Pull request by zygoloid**
When declaring an associated entity in an interface -- just associated functions for now -- create an associated entity value and corresponding type to represent a "slot in a witness table". Also track the list of associated entities on the interface so that we will eventually be able to check impls against them.

Associated entities are represented as the integer index of their slot in a witness table.

### CODE REVIEW

File: toolchain/sem_ir/interface.h [hunk 1]
```diff
@@ -20,6 +20,12 @@ struct Interface : public Printable<Interface> {
   // until we reach the `}` of the interface definition.
   auto is_defined() const -> bool { return defined; }
 
+  // Determines whether we're currently defining the interface. This is true
+  // between the braces of the interface.
+  auto is_being_defined() const -> bool {
```
Comment (jonmeow) [inline toolchain/sem_ir/interface.h:25]: FWIW, I was considering "partially defined" for this state. But I was also mulling doing equivalent for other `definition_id.is_valid()` checks (class and fn), so maybe the `!defined` check is extra we don't generally want? *shrug*

File: toolchain/sem_ir/interface.h [hunk 2]
```diff
@@ -20,6 +20,12 @@ struct Interface : public Printable<Interface> {
...
+    return definition_id.is_valid() && !defined;
```
Comment (jonmeow) [inline toolchain/sem_ir/interface.h]:
```suggestion
    return definition_id.is_valid() && !is_defined();
```

Just to enforce the relationship.

File: toolchain/check/decl_name_stack.cpp [hunk 3]
```diff
@@ -5,7 +5,9 @@
 #include "toolchain/check/decl_name_stack.h"
 
 #include "toolchain/check/context.h"
+#include "toolchain/check/scope_index.h"
```
Comment (jonmeow) [inline toolchain/check/decl_name_stack.cpp]: Mirroring includes from the header?

File: toolchain/lower/handle.cpp [hunk 4]
```diff
@@ -60,6 +60,12 @@ auto HandleAssign(FunctionContext& context, SemIR::InstId /*inst_id*/,
   context.FinishInit(storage_type_id, inst.lhs_id, inst.rhs_id);
 }
 
+auto HandleAssociatedEntity(FunctionContext& context, SemIR::InstId inst_id,
+                            SemIR::AssociatedEntity /*inst*/) -> void {
+  context.SetLocal(inst_id,
+                   llvm::ConstantStruct::getAnon(context.llvm_context(), {}));
```
Comment (jonmeow) [inline toolchain/lower/handle.cpp]: Should there be lowering tests?

File: toolchain/sem_ir/inst.h [hunk 5]
```diff
@@ -343,6 +343,16 @@ class InstStore {
...
   }
 
+  // Returns the requested instruction as the specified type, if it is valid and
+  // of that type. Otherwise returns nullopt.
+  template <typename InstT>
+  auto TryGetAsIfValid(InstId inst_id) const -> std::optional<InstT> {
+    if (!inst_id.is_valid()) {
+      return std::nullopt;
+    }
+    return Get(inst_id).TryAs<InstT>();
```
Comment (jonmeow) [inline toolchain/sem_ir/inst.h]:
```suggestion
    return TryGetAs<InstT>(inst_id);
```

?

File: toolchain/sem_ir/interface.h [hunk 6]
```diff
@@ -42,6 +48,8 @@ struct Interface : public Printable<Interface> {
 
   // The following members are set at the `}` of the interface definition.
   bool defined = false;
+  // The declarations of the associated entities.
+  InstBlockId associated_entities_id = InstBlockId::Invalid;
```
Comment (jonmeow) [inline toolchain/sem_ir/interface.h]: Is this always set? If so, can `defined` be removed? If not, maybe the comment should note it?

File: toolchain/check/testdata/impl/impl_forall.carbon [hunk 7]
```diff
@@ -16,19 +16,23 @@ impl forall [T:! type] T as Simple {
...
 // CHECK:STDOUT:   %.1: type = interface_type @Simple [template]
+// CHECK:STDOUT:   %.2: type = assoc_entity_type @Simple, <function> [template]
+// CHECK:STDOUT:   %.3: <associated <function> in Simple> = assoc_entity element0, @Simple.%F [template]
 // CHECK:STDOUT: }
 // CHECK:STDOUT:
 // CHECK:STDOUT: file {
 // CHECK:STDOUT:   package: <namespace> = namespace {.Simple = %Simple.decl} [template]
 // CHECK:STDOUT:   %Simple.decl = interface_decl @Simple, () [template = constants.%.1]
-// CHECK:STDOUT:   impl_decl @impl, (<unexpected instref inst+4>, <unexpected instref inst+5>, <unexpected instref inst+6>, <unexpected instref inst+7>)
+// CHECK:STDOUT:   impl_decl @impl, (<unexpected instref inst+7>, <unexpected instref inst+8>, <unexpected instref inst+9>, <unexpected instref inst+10>)
```
Comment (jonmeow) [inline toolchain/check/testdata/impl/impl_forall.carbon:26]: I think these "unexpected instref" should be fixed because this is a non-fail test.

**Review by jonmeow — APPROVED** [review]
Please at least fix the "unexpected instref", that looks like it should be fixed here. Other than that, generally looks good, let me know if you want another round of review.

File: toolchain/check/decl_name_stack.cpp (hunk 3 above)
Comment (zygoloid) [reply toolchain/check/decl_name_stack.cpp]: VSCode did it, I dunno why. Undone.

File: toolchain/check/testdata/impl/impl_forall.carbon (hunk 7 above)
Comment (zygoloid) [reply toolchain/check/testdata/impl/impl_forall.carbon:26]: I agree. Will take a look after this lands (they're not new here or related to this change at all).

File: toolchain/lower/handle.cpp (hunk 4 above)
Comment (zygoloid) [reply toolchain/lower/handle.cpp]: Added, and in fact this didn't work :) Fixed.

File: toolchain/sem_ir/interface.h (hunk 6 above)
Comment (zygoloid) [reply toolchain/sem_ir/interface.h]: Sure, done.

**Pull request merged**